### ISSUE TITLE
Using CUDA or OpenCL

### ISSUE THREAD

**Issue by necros2k7** [libjxl/libjxl#1134]
**Is your feature request related to a problem? Please describe.**
Nope

**Describe the solution you'd like**
using CUDA or OPENCL to accelerate encoding/decoding

**Comment by KoromaruKoruko** [comment]
would accelerating encoding/decoding be of any real benefit?

as I can say for non-handcrafted decodes, it's not worth the added effort and may likely end up slowing the process down.

**Comment by ValZapod** [comment]
> would accelerating encoding/decoding be of any real benefit?

Intel HW jpeg decoder is used in some application and is to be supported in Chrome. Same from NVjpeg library.

**Comment by KoromaruKoruko** [comment]
> > would accelerating encoding/decoding be of any real benefit?
> 
> Intel HW jpeg decoder is used in some applications and is to be supported in Chrome. Same from nvJPEG library.

The Intel HW* Jpeg decoder is an example decoder utilizing OpenCL for Huffman and inverse DCT decoding.
The Question is, is it worth the effort to implement support within the reference library. For hardware accelerated applications this 100% would be faster as you can implement it as a one short texture shader. As the issue with the Intel HW* Jpeg decoder is the PCIe bus speed. If we implement this into the reference library. we would 1 need to implement a GPU device and OpenCL Instance manager, 2 transfer the entire image data to and from the GPU for decoding, and 3 heavily modify the library to support OpenCL.
which simply said the transferring of the image to and from GPU memory. could cause slowdowns due to bus saturation, especially if that image is just sent straight back to the GPU to be rendered via HW accelerated renderer.

in short, its not a bad idea, its just not exactly something to implement in reference software. let people who create their own implementation take advantage of their environments.

**Comment by ValZapod** [comment]
BTW, I totally missed this! Wow. https://docs.nvidia.com/cuda/nvjpeg2000/userguide.html

**Comment by ValZapod** [comment]
BTW, 

>The A100 includes a 5-core hardware JPEG decode engine. nvJPEG takes advantage of the hardware backend for batched processing of JPEG images.

So apparently A100 for servers can even decode jpeg literally with HW accel.

**Comment by LifeIsStrange** [comment]
According to https://github.com/elFarto/nvidia-vaapi-driver NVDEC expose the JPEG ASIC

**Comment by ZaquL** [comment]
I just checked ffmpeg.exe -i 02-14817ce690189825ddc36fe847d7f436.png -distance 0 -effort 9 jxq2.jxl

And it is so slow (the default is -effort 7)!! CUDA is really needed. 

Interestingly enough cjxl.exe 02-14817ce690189825ddc36fe847d7f436.png image.jxl --modular --quality 100 --speed 9 -E 3 is even slower but compresses slightly better.

**Comment by RafaelLinux** [comment]
> would accelerating encoding/decoding be of any real benefit?
> 
> as I can say for non-handcrafted decodes, it's not worth the added effort and may likely end up slowing the process down.

Two years have passed and it seems that this functionality has not been considered interesting, when in fact it is very interesting. JXL requires a lot of computation in compression and when it is done massively, the CPU practically "burns". Does nobody really think that it is important to be able to use the GPU to encode in JXL?

**Comment by lantudou** [comment]
Performing JXL encoding and decoding on the GPU is highly advantageous for increasing throughput during AI-powered image inference. Furthermore, the high compression ratio inherent to JXL provides significant bandwidth savings.
